### PR TITLE
Fix valid_courses to give the right information based on user

### PR DESCRIPTION
**Pull request by dmcavoy**
The curriculum team reported that the assign to section button was not working on the Course Overview page for pilot courses. I traced it back to issues with the `assignable_for_user?` path. The user was not getting passed through.

Adding it and updating a couple things in the valid_courses method led to the button showing for pilot courses for pilot teachers and levelbuilders:

Screen Shot 2021-09-16 at 9 27 32 PM

## Links

Jira: https://codedotorg.atlassian.net/browse/PLAT-1309

## Testing story

Added unit tests to prevent issue from happening again

### CODE REVIEW

File: dashboard/app/models/unit_group.rb [hunk 1]
```diff
@@ -312,24 +309,19 @@ def self.has_any_course_experiments?(user)
     Experiment.any_enabled?(user: user, experiment_names: UnitGroupUnit.experiments)
   end
 
-  # Get the set of valid courses for the dropdown in our sections table.
-  def self.valid_courses_without_cache
```
Comment (dmcavoy) [inline dashboard/app/models/unit_group.rb:316]: This does not seem to be used anywhere

Comment (davidsbailey) [reply dashboard/app/models/unit_group.rb:316]: Thank you. it looks like this got orphaned in https://github.com/code-dot-org/code-dot-org/pull/42228.

File: dashboard/app/models/unit_group.rb [hunk 2]
```diff
@@ -312,24 +309,19 @@ def self.has_any_course_experiments?(user)
...
-    UnitGroup.all.select(&:launched?)
-  end
-
   # Returns whether the course id is valid, even if it is not "stable" yet.
   # @param course_id [String] id of the course we're checking the validity of
   # @return [Boolean] Whether this is a valid course ID
   def self.valid_course_id?(course_id, user = nil)
-    valid_courses(user: user).any? {|unit_group| unit_group.id == course_id.to_i}
+    UnitGroup.valid_courses(user: user).any? {|unit_group| unit_group.id == course_id.to_i}
```
Comment (davidsbailey) [inline dashboard/app/models/unit_group.rb:316]: was this change intentional? `UnitGroup.` should be implied when called from another class method.

File: dashboard/app/models/unit_group.rb [hunk 3]
```diff
@@ -312,24 +309,19 @@ def self.has_any_course_experiments?(user)
...
   end
 
   # @param user [User]
   # @returns [Boolean] Whether the user can assign this course.
   # Users should only be able to assign one of their valid courses.
   def assignable_for_user?(user)
     if user&.teacher?
-      UnitGroup.valid_course_id?(id)
+      UnitGroup.valid_course_id?(id, user)
```
Comment (davidsbailey) [inline dashboard/app/models/unit_group.rb:324]: this looks like it removes the last caller of `valid_course_id?` which does not specify a user. what do you think about making `user` a required param to help avoid this kind of problem in the future?

File: dashboard/test/models/unit_group_test.rb [hunk 4]
```diff
@@ -866,6 +866,31 @@ class ProgressTests < ActiveSupport::TestCase
...
+    levelbuilder = create :levelbuilder
+    pilot_teacher = create :teacher, pilot_experiment: 'my-experiment'
+    pilot_course = create :unit_group, pilot_experiment: 'my-experiment'
+    assert UnitGroup.any?(&:pilot?)
+
+    refute pilot_course.assignable_for_user?(student)
+    refute pilot_course.assignable_for_user?(teacher)
+    assert pilot_course.assignable_for_user?(pilot_teacher)
+    assert pilot_course.assignable_for_user?(levelbuilder)
+  end
```
Comment (davidsbailey) [inline dashboard/test/models/unit_group_test.rb:892]: nice tests! this seems like a good method to test, rather than adding a bunch of new tests higher up in the call stack.

**Review by davidsbailey — APPROVED** [review]

File: dashboard/app/models/unit_group.rb (hunk 2 above)
Comment (dmcavoy) [reply dashboard/app/models/unit_group.rb:316]: I think I might have done it when valid_courses still wasn't working because I thought it might be the issue. I can put it back if needed

File: dashboard/app/models/unit_group.rb (hunk 3 above)
Comment (dmcavoy) [reply dashboard/app/models/unit_group.rb:324]: That seems like a good idea. I'll try updating.
